### PR TITLE
Add higher level session (Tunnistamo Session) feature

### PR DESCRIPTION
**Pull request by mikkokeskinen**
This PR adds a feature where social auth logins, OIDC provider tokens, and Django sessions are combined in one higher level session.

The new session is used to get the "loa" value from the social auth instance the user used when creating OIDC tokens. Additionally the "amr" value is now read from the social auth and not from the user.last_login_backend-field. The last_login_backend is not removed because it is still in use elsewhere.

There is a new setting `TRUSTED_LOA_BACKENDS` which should be set to include all the backends whose "loa" value we trust.

This PR doesn't implement the deletion of the sessions or tokens if the user logs out. That functionality should be done when token invalidation is implemented.

Refs HP-647

### CODE REVIEW

File: tunnistamo/tests/conftest.py [hunk 1]
```diff
@@ -1,12 +1,24 @@
...
 from django.utils.crypto import get_random_string
-from oidc_provider.models import Code
+from oidc_provider.models import RESPONSE_TYPE_CHOICES
+from oidc_provider.models import Client as OidcClient
+from oidc_provider.models import Code, ResponseType
 
 from oidc_apis.models import Api, ApiDomain, ApiScope
-from users.tests.conftest import loginmethod_factory, oidcclient_factory, user  # noqa
+from oidc_apis.views import get_api_tokens_view
+from tunnistamo.tests.test_restricted_auth import create_rsa_key
```
Comment (akikoskinen) [inline tunnistamo/tests/conftest.py:17]: Would probably be good to move `create_rsa_key` into this file.

File: tunnistamo/tests/conftest.py [hunk 2]
```diff
@@ -1,12 +1,24 @@
...
+from users.tests.conftest import (  # noqa
+    DummyOidcBackendBase, loginmethod_factory, oidcclient_factory, tunnistamosession_factory, user,
+    usersocialauth_factory
+)
```
Comment (akikoskinen) [inline tunnistamo/tests/conftest.py:21]: My IDE says that of these imports, only `DummyOidcBackendBase` is really used. Why are the others imported?

**Review by akikoskinen — CHANGES_REQUESTED** [review]
Generally looks very solid. Some small observations.

Also go through the Codecov warnings if they indicate any code paths that should be covered by tests.

File: users/tests/conftest.py [hunk 3]
```diff
@@ -183,6 +184,20 @@ def service():
...
 
 
+@pytest.fixture()
+def tunnistamosession_factory():
+    def make_instance(**args):
+        args.setdefault('data', {})
+        args.setdefault('created_at', timezone.now())
+
+        instance = TunnistamoSession.objects.create(**args)
+        instance.save()
```
[the remark this answers is not in the document]
Comment (mikkokeskinen) [reply users/tests/conftest.py]: True. Just copy-pasted the previous style made by ...checks blame... your truly in 2017. 😄

File: users/tests/conftest.py [hunk 4]
```diff
@@ -183,6 +184,20 @@ def service():
     return ServiceFactory(target='client')
 
 
+@pytest.fixture()
+def tunnistamosession_factory():
+    def make_instance(**args):
+        args.setdefault('data', {})
+        args.setdefault('created_at', timezone.now())
```
[the remark this answers is not in the document]
Comment (mikkokeskinen) [reply users/tests/conftest.py]: Same thing as in the previous. I changed them all.

File: users/models.py [hunk 5]
```diff
@@ -3,8 +3,12 @@
 import logging
 import uuid
 
+from django.contrib.contenttypes.fields import GenericForeignKey
+from django.contrib.contenttypes.models import ContentType
 from django.contrib.postgres.fields import JSONField
+from django.core.exceptions import ObjectDoesNotExist
 from django.db import models
+from django.utils import timezone
```
[the remark this answers is not in the document]
Comment (mikkokeskinen) [reply users/models.py]: Done.

File: tunnistamo/tests/conftest.py [hunk 6]
```diff
@@ -74,3 +86,228 @@ def make_instance(**args):
...
+    if trust_loa:
+        settings.TRUSTED_LOA_BACKENDS = [
+            DummyFixedOidcBackend.name
+        ]
+
+    if not test_client:
+        test_client = TestClient()
+
+    # Import in method to prevent circular import
+    from auth_backends.tests.test_oidc_backchannel_logout import reload_social_django_utils
```
[the remark this answers is not in the document]
Comment (mikkokeskinen) [reply tunnistamo/tests/conftest.py]: Done.

File: users/tests/test_tunnistamo_session.py [hunk 7]
```diff
@@ -0,0 +1,321 @@
...
+    tunnistamo_session.add_element(social_auth)
+
+    social_auth.delete()
+
+    assert tunnistamo_session.elements.count() == 1
+    assert tunnistamo_session.get_content_object_by_model(UserSocialAuth) is None
+
+
+@pytest.mark.django_db
+def test_manager_get_or_create_from_request_no_user_in_request(rf):
```
[the remark this answers is not in the document]
Comment (mikkokeskinen) [reply users/tests/test_tunnistamo_session.py]: Done.

File: tunnistamo/tests/conftest.py (hunk 1 above)
Comment (mikkokeskinen) [reply tunnistamo/tests/conftest.py]: Done.

File: tunnistamo/tests/conftest.py (hunk 2 above)
Comment (mikkokeskinen) [reply tunnistamo/tests/conftest.py]: (Some of) Those are used in the test files. i.e. imported fixtures from another conftest. Removed the unused ones.

**Comment by mikkokeskinen** [comment]
> The commit message mentions "TunnistamoAuthorizeEndpoint" but I suppose it should say "TunnistamoTokenEndpoint".

True. Changed.

**Review by akikoskinen — DISMISSED** [review]

**Review by tituomin — DISMISSED** [review]

**Comment by akikoskinen** [comment]
While the new tests thoroughly test the `get_loa` function, I wonder if it would be better to have "full stack" tests for this? For example, are there tests ensuring that a `social.extra_data['id_token']` is added by some party in the first place, so that it can be used in the `get_loa` function?

**Comment by mikkokeskinen** [comment]
> While the new tests thoroughly test the `get_loa` function, I wonder if it would be better to have "full stack" tests for this? For example, are there tests ensuring that a `social.extra_data['id_token']` is added by some party in the first place, so that it can be used in the `get_loa` function?

Many of the tests are already "end to end" tests. e.g. https://github.com/City-of-Helsinki/tunnistamo/pull/169/commits/80d0aa1a5f331ed1954c2570bb52ca4e7324d03b#diff-e5259a6b6234cd6f3ceb1adeea84ede7013f35f92a1a108c750b4e76bc750393R37

They use the dummy provider though, but I don't see it matters.

**Comment by akikoskinen** [comment]
> Many of the tests are already "end to end" tests. e.g. [80d0aa1#diff-e5259a6b6234cd6f3ceb1adeea84ede7013f35f92a1a108c750b4e76bc750393R37](https://github.com/City-of-Helsinki/tunnistamo/commit/80d0aa1a5f331ed1954c2570bb52ca4e7324d03b#diff-e5259a6b6234cd6f3ceb1adeea84ede7013f35f92a1a108c750b4e76bc750393R37)

That's good. But I was wondering if it needs to be tested that the expected `social.extra_data['id_token']` value really is put into place by some party (I don't know if this _party_ is some library or Tunnistamo code, so I'm referring to it in this abstract way), so that it can be read. For example, what if the party  actually puts the data into `social.extra_datas['id_token']`, but both the `get_loa` function and current test code expect another attribute name? It would look like fully tested, but as a whole the system wouldn't work.

**Comment by mikkokeskinen** [comment]
Turns out the id_token-attribute is just used temporarily by the Open ID connect backend. So the proper attribute is the extra_data. Which in turn is filled by the social auth library.

**Comment by akikoskinen** [comment]
Only now I really realized that the backend used in tests (`DummyFixedOidcBackend`) has its own implementation of `get_loa`. The implementation is currently identical to that in `HelsinkiTunnistus`, but do the tests now depend on these implementations also **staying** identical? Would it be possible to refactor the code (e.g. extract a helper function or mixin or something else) in a way that the same implementation would be used both in production and test code?

**Comment by mikkokeskinen** [comment]
I thought about it, but it would be yet another abstraction to decipher when reading the backend implementation. I don't see them diverting as a risk as the helsinkitunnistussuomifi-backend has its own test for the get_loa-method.

**Review by akikoskinen — APPROVED** [review]
I guess all good then.

**Pull request merged**